chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -16,30 +16,91 @@
 use static_assertions::assert_eq_size;
 use std::collections::BTreeMap;
 
-use crate::{error::ConnectTransactionError, TransactionSource};
+use crate::error::ConnectTransactionError;
 
 use common::{
     chain::{Block, DelegationId, PoolId},
     primitives::{per_thousand::PerThousand, Amount, Id},
     Uint256,
 };
-use pos_accounting::{
-    AccountingBlockRewardUndo, PoSAccountingOperations, PoSAccountingUndo, PoSAccountingView,
-};
+use pos_accounting::{PoSAccountingOperations, PoSAccountingUndo, PoSAccountingView, PoolData};
 use utils::ensure;
 
-use super::pos_accounting_delta_adapter::PoSAccountingDeltaAdapter;
+pub trait DelegationSharesView {
+    type Error: std::error::Error;
+
+    fn find_pool_data(&self, pool_id: PoolId) -> Result<Option<PoolData>, Self::Error>;
+
+    fn get_pool_delegations_shares(
+        &self,
+        pool_id: PoolId,
+    ) -> Result<Option<BTreeMap<DelegationId, Amount>>, Self::Error>;
+}
+
+impl<T> DelegationSharesView for T
+where
+    T: PoSAccountingView,
+{
+    type Error = T::Error;
+    fn find_pool_data(&self, pool_id: PoolId) -> Result<Option<PoolData>, Self::Error> {
+        self.get_pool_data(pool_id)
+    }
+
+    fn get_pool_delegations_shares(
+        &self,
+        pool_id: PoolId,
+    ) -> Result<Option<BTreeMap<DelegationId, Amount>>, Self::Error> {
+        self.get_pool_delegations_shares(pool_id)
+    }
+}
+
+pub trait DelegationSharesOperations<U> {
+    fn add_reward_to_pool_pledge(
+        &mut self,
+        pool_id: PoolId,
+        amount_to_add: Amount,
+    ) -> Result<U, pos_accounting::Error>;
+
+    fn add_reward_to_delegate(
+        &mut self,
+        delegation_target: DelegationId,
+        amount_to_delegate: Amount,
+    ) -> Result<U, pos_accounting::Error>;
+}
+
+impl<T> DelegationSharesOperations<PoSAccountingUndo> for T
+where
+    T: PoSAccountingOperations,
+{
+    fn add_reward_to_pool_pledge(
+        &mut self,
+        pool_id: PoolId,
+        amount_to_add: Amount,
+    ) -> Result<PoSAccountingUndo, pos_accounting::Error> {
+        self.increase_pool_pledge_amount(pool_id, amount_to_add)
+    }
+
+    fn add_reward_to_delegate(
+        &mut self,
+        delegation_target: DelegationId,
+        amount_to_delegate: Amount,
+    ) -> Result<PoSAccountingUndo, pos_accounting::Error> {
+        self.delegate_staking(delegation_target, amount_to_delegate)
+    }
+}
 
 /// Distribute reward among the pool's owner and delegations
-pub fn distribute_pos_reward<P: PoSAccountingView>(
-    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+pub fn distribute_pos_reward<
+    U,
+    P: DelegationSharesView<Error = pos_accounting::Error> + DelegationSharesOperations<U>,
+>(
+    accounting_adapter: &mut P,
     block_id: Id<Block>,
     pool_id: PoolId,
     total_reward: Amount,
-) -> Result<AccountingBlockRewardUndo, ConnectTransactionError> {
+) -> Result<Vec<U>, ConnectTransactionError> {
```

Comment:
Isn't the returning type always `PoSAccountingUndo`?

Reply:
not for the scanner, it doesn't need the undos

Reply:
right I see now